indexer/categories_holder.cpp

```diff
@@ -55,7 +55,7 @@ bool ParseEmoji(CategoriesHolder::Category::Name & name)
 
 void FillPrefixLengthToSuggest(CategoriesHolder::Category::Name & name)
 {
-  if (std::isdigit(name.m_name.front()) && name.m_name.front() != '0')
+  if (std::isdigit(static_cast<unsigned char>(name.m_name.front())) && name.m_name.front() != '0')
```

Comment:
@Osyotr 
Those casts to unsigned char break the tests on Mac. They hang on parsing the arabic UTF-8 string by not seeing a valid UTF-8 character.

These functions should take a signed `int` type instead of unsigned.

Reply:
Thanks for investigating.
Cppreference explicitly mentions that passing something that is not representable as `unsigned char` is wrong.
https://en.cppreference.com/w/cpp/string/byte/isdigit
I'll look further into it.

Reply:
The function signature has accepted int since ancient C times: `int isdigit( int ch );` so using int should be enough.

Yes, it is interesting to understand why it hangs.

This code returned 0:
```
    template <typename octet_iterator>
    inline typename std::iterator_traits<octet_iterator>::difference_type
    sequence_length(octet_iterator lead_it)
    {
        uint8_t lead = utf8::internal::mask8(*lead_it);
        if (lead < 0x80)
            return 1;
        else if ((lead >> 5) == 0x6)
            return 2;
        else if ((lead >> 4) == 0xe)
            return 3;
        else if ((lead >> 3) == 0x1e)
            return 4;
        else
            return 0;
    }
```

So this loop in `Move` was indefinite:
```
    while (m_end != m_finish && !m_delimFn(*m_end))
      ++m_end;
```

It is called when search categories are loaded in `void CategoriesHolder::LoadFromStream(std::istream & s)`

Reply:
As we're using unsafe utf8 iteration without checks/crashes in cases of invalid text, somehow iterator pointed to a char value in the string that could not be incremented.

Reply:
@Osyotr does it make sense to change the code to `int`, merge PR, and proceed with the investigation separately?

Reply:
Makes sense to move this commit to another PR. I don't think the change is wrong.
I'll do it later today (or your can do it now since you've got write access to this PR).

Reply:
Hanging code on Mac either proves that it's wrong or hints about a bug in the recent Mac OS clang.

Any help with PRs is greatly appreciated 🙌

Reply:
@Osyotr the failing line contains arabic chars, I see this in debugger:
```
"ar:طعام|أمكان لتناول الطعا\xd9"
```
It's line 80 from categories.txt
```
ar:طعام|أمكان لتناول الطعام
```

m_start says zero, m_end is strange
<img width="985" alt="выява" src="https://github.com/organicmaps/organicmaps/assets/170263/cfac38e1-5c32-48cb-bfcf-16547248f26f">